python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,623 @@
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_max(
+  T const* data, int const items_per_thread, size_type size, T init_val, T* smax)
+{
+  T local_max = init_val;
+
+// Calculate local max for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load    = data[threadIdx.x + item * blockDim.x];
+      local_max = max(local_max, load);
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local max for each group
+  atomicMax(smax, local_max);
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_min(
+  T const* data, int const items_per_thread, size_type size, T init_val, T* smin)
+{
+  T local_min = init_val;
+
+// Calculate local min for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load    = data[threadIdx.x + item * blockDim.x];
+      local_min = min(local_min, load);
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local min for each group
+  atomicMin(smin, local_min);
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_idxmax(T const* data,
+                              int const items_per_thread,
+                              int64_t const* index,
+                              size_type size,
+                              T init_val,
+                              T* smax,
+                              int64_t* sidx)
+{
+  // Calculate how many elements each thread is working on
+  T local_max       = init_val;
+  int64_t local_idx = -1;
+
+// Calculate local max for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load = data[threadIdx.x + item * blockDim.x];
+      if (load > local_max) {
+        local_max = load;
+        local_idx = index[threadIdx.x + item * blockDim.x];
+      }
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local max for each group
+  atomicMax(smax, local_max);
+
+  __syncthreads();
+
+  if (local_max == (*smax)) { atomicMin(sidx, local_idx); }
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_idxmin(T const* data,
+                              int const items_per_thread,
+                              int64_t const* index,
+                              size_type size,
+                              T init_val,
+                              T* smin,
+                              int64_t* sidx)
+{
+  T local_min       = init_val;
+  int64_t local_idx = -1;
+
+// Calculate local max for each thread
```

Comment:
Incorrect comment, also, all these comments don't say anything the code doesn't say, so I would just remove them.